zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java

```diff
@@ -522,23 +525,24 @@ private void tryRejectingIfUserCommand(final String errorMessage) {
     final var rejectionReason = errorMessage != null ? errorMessage : "";
     final ProcessingResultBuilder processingResultBuilder =
         new BufferedProcessingResultBuilder(logStreamWriter::canWriteEvents);
-    // reset value to minimize any potential error loop that can be caused by writing the full
-    // record.
-    typedCommand.getValue().reset();
-    final var rejectionMetadata =
+    final var errorRecord = new ErrorRecord();
+    errorRecord.initErrorRecord(new RuntimeException(rejectionReason), currentRecord.getPosition());
+
+    final var recordMetadata =
         new RecordMetadata()
-            .recordType(RecordType.COMMAND_REJECTION)
-            .intent(typedCommand.getIntent())
-            .rejectionType(RejectionType.PROCESSING_ERROR)
-            .rejectionReason(rejectionReason);
-    processingResultBuilder.appendRecord(
-        currentRecord.getKey(), typedCommand.getValue(), rejectionMetadata);
+            .recordType(RecordType.EVENT)
+            .valueType(ValueType.ERROR)
+            .intent(ErrorIntent.CREATED)
+            .recordVersion(RecordMetadata.DEFAULT_RECORD_VERSION)
+            .rejectionType(RejectionType.NULL_VAL)
+            .rejectionReason("");
+    processingResultBuilder.appendRecord(currentRecord.getKey(), errorRecord, recordMetadata);
     processingResultBuilder.withResponse(
         RecordType.COMMAND_REJECTION,
         typedCommand.getKey(),
         typedCommand.getIntent(),
-        typedCommand.getValue(),
-        typedCommand.getValueType(),
+        errorRecord,
+        ValueType.ERROR,
```

Comment:
:question: Here I'm undecided whether to send the error record or a null. The problem with null is, it may break several parts because we assume a record always have a non-null value. The value is ignored by the gateway anyway if it is a rejection.

Reply:
I think that's perfectly fine. It only gets serialized at the transport layer, so we could still decide there to drop it.